src/node/configuration_tracker.h

```diff
@@ -0,0 +1,413 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "ccf/app_interface.h"
+#include "ccf/entity_id.h"
+#include "ccf/tx_id.h"
+#include "consensus/aft/raft_types.h"
+#include "ds/logger.h"
+#include "enclave/rpc_sessions.h"
+#include "kv/kv_types.h"
+#include "kv/store.h"
+#include "node/config.h"
+#include "node/entities.h"
+#include "node/nodes.h"
+#include "node/rpc/node_call_types.h"
+#include "node/rpc/serdes.h"
+#include "rpc/serialization.h"
+
+using namespace ccf;
+
+namespace aft
+{
+  using Index = uint64_t;
+
+  struct NodeState
+  {
+    kv::Configuration::NodeInfo node_info;
+
+    // the highest index sent to the node
+    Index sent_idx;
+
+    // the highest matching index with the node that was confirmed
+    Index match_idx;
+
+    // Flag indicating whether this node is still catching up; before that it's
+    // not an eligible voter.
+    bool catching_up;
+
+    NodeState() = default;
+
+    NodeState(
+      const kv::Configuration::NodeInfo& node_info_,
+      Index sent_idx_,
+      Index match_idx_ = 0,
+      bool catching_up_ = true) :
+      node_info(node_info_),
+      sent_idx(sent_idx_),
+      match_idx(match_idx_),
+      catching_up(catching_up_)
+    {}
+  };
+
+  class ConfigurationTracker
+  {
+  public:
+    const NodeId& node_id;
+    ConsensusType consensus_type;
+    std::shared_ptr<kv::AbstractStore> store;
+    std::shared_ptr<enclave::RPCSessions> rpcsessions;
+    std::shared_ptr<enclave::RPCMap> rpc_map;
+    crypto::KeyPairPtr node_sign_kp;
+    const crypto::Pem& node_cert;
+    std::unordered_map<ccf::NodeId, aft::NodeState>& nodes;
+    std::list<kv::Configuration> configurations;
```

Comment:
I don't think `node_id` and `node_cert` should be references, they should just be copied locally (they may be referring to a heap-allocated object's members right now, but they could easily come from the stack and this would break). `nodes` I don't know, that may really be referring to an map owned by someone else? But that should then be some kind of smart pointer, not a reference.

Reply:
The whole class is just a mental separation of the configuration-relevant parts of `Aft`. It's not helpful though, so I'll move all of this back into `Aft`.